File: solutions/leetcode/dota2-senate/main.cpp
class Solution {
   public:
    string predictPartyVictory(string senate) {
        int dCount = 0, dBan = 0;
        int rCount = 0, rBan = 0;

        queue<char> q;
        for (char c : senate) {
            q.push(c);
            if (c == 'R') {
                ++rCount;
            } else {
                ++dCount;
            }
        }

        while (rCount > 0 && dCount > 0) {
            char c = q.front();
            q.pop();

            if (c == 'R') {
                if (rBan > 0) {
                    --rBan;
                    --rCount;
                } else {
                    ++dBan;
                    q.push('R');
                }
            } else {
                if (dBan > 0) {
                    --dBan;
                    --dCount;
                } else {
                    ++rBan;
                    q.push('D');
                }
            }
        }

        return rCount > 0 ? "Radiant" : "Dire";
    }
};

class Solution {
   public:
    string predictPartyVictory(string senate) {
        int rCount = 0, dCount = 0;
        int rBan = 0, dBan = 0;

        queue<char> q;
        for (char c : senate) {
            c == 'R' ? ++rCount : ++dCount;
            q.push(c);
        }

        while (rCount > 0 && dCount > 0) {
            char c = q.front();
            q.pop();

            if (c == 'R') {
                if (dBan > 0) {
                    --dBan, --rCount;
                    continue;
                }
                ++rBan;
            } else {  // c == 'D'
                if (rBan > 0) {
                    --rBan, --dCount;
                    continue;
                }
                ++dBan;
            }

            q.push(c);
        }

        return rCount > 0 ? "Radiant" : "Dire";
    }
};

File: solutions/leetcode/dota2-senate/main.go
func predictPartyVictory(senate string) string {
	radiant := 0
	dire := 0
	q := []rune{}
	for _, c := range senate {
		if c == 'R' {
			radiant++
		} else { // c == 'D'
			dire++
		}
		q = append(q, c)
	}

	r := 0
	d := 0
	for radiant > 0 && dire > 0 {
		c := q[0]
		q = q[1:]
		if c == 'R' {
			if r < 0 {
				r++
				radiant--
			} else {
				d--
				q = append(q, c)
			}
		} else { // c == 'D'
			if d < 0 {
				d++
				dire--
			} else {
				r--
				q = append(q, c)
			}
		}
	}

	if radiant > 0 {
		return "Radiant"
	}
	return "Dire"
}
